core/src/test/scala/akka/persistence/jdbc/state/JdbcDurableStateSpec.scala

```diff
@@ -0,0 +1,181 @@
+    extends PluginSpec(config)
+    with AnyWordSpecLike
+    with BeforeAndAfterAll
+    with BeforeAndAfterEach
+    with Matchers
+    with ScalaFutures
+    with DropCreate {
+
+  implicit lazy val ec = system.dispatcher
+
+  val customSerializers = ConfigFactory.parseString("""
+      akka.actor {
+        serializers {
+          my-payload = "akka.persistence.jdbc.state.MyPayloadSerializer"
+        }
+        serialization-bindings {
+          "akka.persistence.jdbc.state.MyPayload" = my-payload
+        }
+      }
+    """)
+
+  lazy val cfg = system.settings.config
+    .getConfig("jdbc-durable-state-store")
+    .withFallback(config.getConfig("jdbc-durable-state-store"))
+    .withFallback(customSerializers.getConfig("akka.actor"))
+
+  lazy val db = SlickDatabase.database(cfg, new SlickConfiguration(cfg.getConfig("slick")), "slick.db")
+  lazy val durableStateConfig = new DurableStateTableConfiguration(cfg)
+  implicit lazy val system: ActorSystem = ActorSystem("JdbcDurableStateSpec", config.withFallback(customSerializers))
+  lazy val serialization = SerializationExtension(system)
+
+  val stateStoreString: JdbcDurableStateStore[String]
+  val stateStorePayload: JdbcDurableStateStore[MyPayload]
+
+  override def beforeAll(): Unit = {
+    dropAndCreate(schemaType)
+    super.beforeAll()
+  }
+
+  override def beforeEach(): Unit = {
+    dropAndCreate(schemaType)
+    super.beforeAll()
+  }
+
+  override def afterAll(): Unit = {
+    db.close()
+  }
+
+  "A durable state store" must {
+    "not load a state given an invalid persistenceId" in {
+      whenReady {
+        stateStoreString.getObject("InvalidPersistenceId")
+      } { v =>
+        v.value shouldBe None
+      }
+    }
+    "add a valid state successfully" in {
+      whenReady {
+        stateStoreString.upsertObject("p123", 1, "a valid string", "t123")
+      } { v =>
+        v shouldBe akka.Done
+      }
+    }
+    "support composite upsert-fetch-repeat loop" in {
+      whenReady {
+        for {
+
+          n <- stateStoreString.upsertObject("p234", 1, "a valid string", "t123")
+          _ = n shouldBe akka.Done
+          g <- stateStoreString.getObject("p234")
+          _ = g.value shouldBe Some("a valid string")
+          u <- stateStoreString.upsertObject("p234", 2, "updated valid string", "t123")
+          _ = u shouldBe akka.Done
+          h <- stateStoreString.getObject("p234")
+
+        } yield h
+      } { v =>
+        v.value shouldBe Some("updated valid string")
+      }
+    }
+    "fail inserting incorrect sequence number" in {
+      whenReady {
+        (for {
+
+          n <- stateStoreString.upsertObject("p345", 1, "a valid string", "t123")
+          _ = n shouldBe akka.Done
+          g <- stateStoreString.getObject("p345")
+          _ = g.value shouldBe Some("a valid string")
+          u <- stateStoreString.upsertObject("p345", 1, "updated valid string", "t123")
```

Comment:
good, you can also add another test that tries to upsert seqNr 2 twice, which should fail

Reply:
done ..